service/raft/raft_address_map.hh

```diff
@@ -387,6 +388,64 @@ public:
         }
         return *it;
     }
+
+    ///////////////// i_endpoint_state_change_subscriber ///////////////////////////////
+    //
+    future<> on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state ep_state) {
+        auto app_state_ptr = ep_state.get_application_state_ptr(gms::application_state::RAFT_SERVER_ID);
+        if (app_state_ptr) {
+            raft::server_id id(utils::UUID(app_state_ptr->value));
+            add_or_update_entry(id, endpoint);
+        }
+        return make_ready_future<>();
+    }
+
+    virtual future<> on_join(gms::inet_address endpoint, gms::endpoint_state ep_state) {
+
+        rslog.debug("raft_address_map::on_join {} {}", endpoint, ep_state);
+        return on_endpoint_change(endpoint, ep_state);
+    }
+
+    virtual future<> before_change(gms::inet_address endpoint,
+        gms::endpoint_state current_state, gms::application_state new_statekey,
+        const gms::versioned_value& newvalue)  {
+        // Raft server ID never changes - do nothing
+        rslog.debug("raft_address_map::before_change {} {} {}", endpoint, current_state, new_statekey);
+        return make_ready_future<>();
+    }
+
+    virtual future<> on_change(gms::inet_address endpoint, gms::application_state state,
+        const gms::versioned_value& value) {
+        // Raft server ID never changes - do nothing
```

Comment:
on_change() is never called for server id change. The node restarts and begins gossiping its new id. It's not an on_change() event, which happens when a live node changes some of its app state.

Reply:
Want happens in the following scenario:
  1. node A joins (endpoint = 10.0.0.1)
  1. node B joins (endpoint = 10.0.0.2)
  1. node B changes IP to 10.0.0.3
  1. node A changes IP to 10.0.0.2

Suppose that node C gets a single gossip notification for the last two IP changes. Will gossip report this as 10.0.0.2 changing server id from B to A?

That would be incorrect, it's not a change of properties of a single endpoint. 

Perhaps gossiper should be changed to track endpoints by server id.

Reply:
Gossiper never reports IP changes as "on_change" events. The new node comes up with a new IP (on_join event) and the old node just dies and is removed on some point (on_dead event). on_change events are for changes in a state of a normal node. We don't care about on_dead event, since we have raft configuration changes and removal from the configuration. We also don't care much about gossip, because we learn the address map information from incoming Raft RPCs as well, so gossip is just one source of information. (And yes, I complained earlier that this will make debugging this design quite tricky).

There are other issues with your example. Why you believe that it's safe to concurrently change addresses of a living cluster, multiple nodes? Why do you think it's safe to give node A an old address of node B without waiting for the gossip to settle with the previous change first? I don't know if it works at all and how, but I would guess concurrent changes are only reliable if use a linearizable storage, not an eventually consistent one. I complained about that as well when this design was suggested, but was hushed with "this is not Raft concern".

How many notifications will node C get? I presume it shouldn't matter *how many* notifications will arrive, as soon as any notification arrives with a new versioned value(s), we will notice both changes and invoke callbacks on them.

Should gossiper be changed to track endpoints by server id? I don't think gossiper should be tracking or disseminating any state except node liveness, and node liveness should be associated with Raft server id, yes. However, if you look at the earlier Kamil's comment, he would like to avoid adding Raft server id to gossip unless upgrade is finished. I imagine it would be quite tricky to track the state based on ip_addr or or raft server id depending on the state of upgrade or recovery.

Reply:
Resolving as there were no further comments on the explanation above.